rpc/client/http/rpc_test.go

```diff
@@ -0,0 +1,211 @@
+	require.Equal(t, 1, len(vals.Validators))
+	val := vals.Validators[0]
+
+	// make sure the current set is also the genesis set
+	assert.Equal(t, gval.Amount, val.VotingPower)
+	assert.Equal(t, gval.PubKey, val.PubKey)
+}
+
+// Make some app checks
+func TestAppCalls(t *testing.T) {
+	assert, require := assert.New(t), require.New(t)
+	c := GetClient()
+
+	// get an offset of height to avoid racing and guessing
+	s, err := c.Status()
+	require.Nil(err)
+	// sh is start height or status height
+	sh := s.LatestBlockHeight
+
+	// look for the future
+	_, err = c.Block(sh + 2)
+	assert.NotNil(err) // no block yet
+
+	// write something
+	k, v, tx := MakeTxKV()
+	_, err = c.BroadcastTxCommit(tx)
+	require.Nil(err, "%+v", err)
+	// wait before querying
+	time.Sleep(time.Second * 1)
+	qres, err := c.ABCIQuery("/key", k, false)
+	if assert.Nil(err) && assert.True(qres.Response.Code.IsOK()) {
+		data := qres.Response
+		// assert.Equal(k, data.GetKey())  // only returned for proofs
+		assert.Equal(v, data.GetValue())
+	}
+	// +/- 1 making my head hurt
+	h := int(qres.Response.Height) - 1
+
+	// and we can even check the block is added
+	block, err := c.Block(h)
+	require.Nil(err, "%+v", err)
+	appHash := block.BlockMeta.Header.AppHash
+	assert.True(len(appHash) > 0)
+	assert.EqualValues(h, block.BlockMeta.Header.Height)
+
+	// check blockchain info, now that we know there is info
+	// TODO: is this commented somewhere that they are returned
+	// in order of descending height???
+	info, err := c.BlockchainInfo(h-2, h)
+	require.Nil(err, "%+v", err)
+	assert.True(info.LastHeight > 2)
+	if assert.Equal(3, len(info.BlockMetas)) {
+		lastMeta := info.BlockMetas[0]
+		assert.EqualValues(h, lastMeta.Header.Height)
+		bMeta := block.BlockMeta
+		assert.Equal(bMeta.Header.AppHash, lastMeta.Header.AppHash)
+		assert.Equal(bMeta.BlockID, lastMeta.BlockID)
+	}
+
+	// and get the corresponding commit with the same apphash
+	commit, err := c.Commit(h)
+	require.Nil(err, "%+v", err)
+	cappHash := commit.Header.AppHash
+	assert.Equal(appHash, cappHash)
+	assert.NotNil(commit.Commit)
+
+	// compare the commits (note Commit(2) has commit from Block(3))
+	commit2, err := c.Commit(h - 1)
+	require.Nil(err, "%+v", err)
+	assert.Equal(block.Block.LastCommit, commit2.Commit)
+
+	// and we got a proof that works!
+	pres, err := c.ABCIQuery("/key", k, true)
+	if assert.Nil(err) && assert.True(pres.Response.Code.IsOK()) {
+		proof, err := merkle.ReadProof(pres.Response.GetProof())
+		if assert.Nil(err) {
+			key := pres.Response.GetKey()
+			value := pres.Response.GetValue()
+			assert.Equal(appHash, proof.RootHash)
+			valid := proof.Verify(key, value, appHash)
+			assert.True(valid)
+		}
+	}
+}
+
+func TestSubscriptions(t *testing.T) {
+	require := require.New(t)
+	c := GetClient()
+	err := c.StartWebsocket()
+	require.Nil(err)
+	defer c.StopWebsocket()
+
+	// subscribe to a transaction event
+	_, _, tx := MakeTxKV()
+	// this causes a panic in tendermint core!!!
```

Comment:
old comment... I think I pushed a fix to tendermint in October, before I got hired even.

this client code was born out of a personal project.  But yeah, due to some typos and crazy interface magic, subscribing to a tx event would cause tendermint to panic:
https://github.com/tendermint/tendermint/issues/285

Removing this comment now.